File: machine/cpp/input/pointer.cpp
/*
 * pointer.cpp - Pointer input handler implementation
 */

#include "pointer.h"
#include "manager.h"
#include "core/machine_manager.h"
#include <algorithm>
#include <cmath>

namespace bmsx {
namespace {

constexpr const char* kPointerPosition = "pointer_position";
constexpr const char* kPointerDelta = "pointer_delta";
constexpr const char* kPointerWheel = "pointer_wheel";

constexpr const char* kPointerDefaultCodes[] = {
	"pointer_primary",
	"pointer_secondary",
	"pointer_aux",
	"pointer_back",
	"pointer_forward",
	kPointerPosition,
	kPointerDelta,
	kPointerWheel,
};

i32 pointerButtonBit(const std::string& code) {
	if (code == "pointer_primary") return INP_POINTER_BUTTON_PRIMARY;
	if (code == "pointer_aux") return INP_POINTER_BUTTON_AUX;
	if (code == "pointer_secondary") return INP_POINTER_BUTTON_SECONDARY;
	if (code == "pointer_back") return INP_POINTER_BUTTON_BACK;
	if (code == "pointer_forward") return INP_POINTER_BUTTON_FORWARD;
	return -1;
}

}

PointerInput::PointerInput(const std::string& deviceId)
	: m_deviceId(deviceId) {
	reset();
}

void PointerInput::pollInput() {
	const f64 now = MachineManager::instance().clock()->now();
	const f64 prevPollTime = m_lastPollTimeMs;
	m_lastPollTimeMs = now;

	for (auto& [key, state] : m_buttonStates) {
		if (state.pressed) {
			const f64 pressedAt = buttonPressedAtOr(state, now);
			state.presstime = std::max(0.0, now - pressedAt);
			if (prevPollTime > 0.0 && state.justpressed && buttonTimestampOr(state, 0.0) <= prevPollTime) {
				state.justpressed = false;
			}
		} else {
			state.presstime = std::nullopt;
			if (prevPollTime > 0.0 && state.justreleased && buttonTimestampOr(state, 0.0) <= prevPollTime) {
				state.justreleased = false;
			}
		}

		if (key == kPointerDelta) {
			const f64 timestamp = buttonTimestampOr(state, 0.0);
			if (timestamp == m_lastDeltaTimestamp) {
				state.value2d = Vec2{0.0f, 0.0f};
				state.value = 0.0f;
				state.pressed = false;
				state.justpressed = false;
				state.justreleased = false;
			} else {
				m_lastDeltaTimestamp = timestamp;
			}
		} else if (key == kPointerWheel) {
			const f64 timestamp = buttonTimestampOr(state, 0.0);
			if (timestamp == m_lastWheelTimestamp) {
				state.justreleased = state.pressed;
				state.value = 0.0f;
				m_inputControllerWheel = 0.0F;
				state.pressed = false;
				state.justpressed = false;
			} else {
				m_lastWheelTimestamp = timestamp;
				state.justreleased = false;
			}
		}

		state.waspressed = state.waspressed || state.pressed;
		state.wasreleased = state.wasreleased || !state.pressed;
	}
}

ButtonState PointerInput::getButtonState(const ButtonId& button) {
	auto it = m_buttonStates.find(button);
	if (it == m_buttonStates.end()) {
		return ButtonState{};
	}
	return it->second;
}

void PointerInput::consumeButton(const ButtonId& button) {
	auto it = m_buttonStates.find(button);
	if (it == m_buttonStates.end()) {
		return;
	}
	it->second.consumed = true;
	if (button == kPointerWheel) {
		it->second.pressed = false;
		it->second.justpressed = false;
		it->second.justreleased = false;
	}
}

void PointerInput::reset(const std::vector<std::string>* except) {
	if (!except) {
		m_buttonStates.clear();
		for (const char* code : kPointerDefaultCodes) {
			m_buttonStates[code] = ButtonState{};
		}
		m_nextPressId = 1;
		m_lastPositionX = 0.0f;
		m_lastPositionY = 0.0f;
		m_lastPositionValid = false;
		m_lastPollTimeMs = 0.0;
		m_lastDeltaTimestamp = 0.0;
		m_lastWheelTimestamp = 0.0;
		m_inputControllerButtons = 0u;
		m_inputControllerX = 0.0F;
		m_inputControllerY = 0.0F;
		m_inputControllerWheel = 0.0F;
		return;
	}
	resetObject(m_buttonStates, except);
	m_inputControllerButtons = 0u;
	for (const auto& [code, state] : m_buttonStates) {
		if (state.pressed) {
			const i32 bit = pointerButtonBit(code);
			if (bit >= 0) {
				m_inputControllerButtons |= 1u << static_cast<u32>(bit);
			}
		}
	}
	const auto position = m_buttonStates.find(kPointerPosition);
	if (position != m_buttonStates.end() && position->second.value2d.has_value()) {
		m_inputControllerX = position->second.value2d->x;
		m_inputControllerY = position->second.value2d->y;
	} else {
		m_inputControllerX = 0.0F;
		m_inputControllerY = 0.0F;
	}
	const auto wheel = m_buttonStates.find(kPointerWheel);
	m_inputControllerWheel = wheel != m_buttonStates.end() ? wheel->second.value : 0.0F;
}

void PointerInput::ingestButton(const std::string& code, bool down, f32 value,
									f64 timestamp, std::optional<i32> pressId) {
	auto& state = m_buttonStates[code];
	if (down) {
		const i32 resolvedPressId = resolveButtonPressId(pressId, state, m_nextPressId);
		state.pressed = true;
		state.justpressed = true;
		state.justreleased = false;
		state.waspressed = true;
		state.timestamp = timestamp;
		state.pressedAtMs = timestamp;
		state.releasedAtMs = std::nullopt;
		state.value = value;
		state.pressId = resolvedPressId;
		state.consumed = false;
	} else {
		const bool wasPressed = state.pressed;
		state.pressed = false;
		state.justpressed = false;
		state.justreleased = wasPressed;
		state.timestamp = timestamp;
		state.pressedAtMs = std::nullopt;
		state.releasedAtMs = timestamp;
		state.value = 0.0f;
		state.waspressed = state.waspressed || wasPressed;
		state.wasreleased = state.wasreleased || wasPressed;
		if (pressId.has_value()) {
			state.pressId = pressId;
		}
		state.consumed = false;
	}
	const i32 bit = pointerButtonBit(code);
	if (bit >= 0) {
		const u32 mask = 1u << static_cast<u32>(bit);
		m_inputControllerButtons = down ? (m_inputControllerButtons | mask) : (m_inputControllerButtons & ~mask);
	}
}

void PointerInput::ingestAxis2(const std::string& code, f32 x, f32 y, f64 timestamp) {
	auto& state = m_buttonStates[code];
	const f32 dx = m_lastPositionValid ? (x - m_lastPositionX) : 0.0f;
	const f32 dy = m_lastPositionValid ? (y - m_lastPositionY) : 0.0f;
	m_lastPositionX = x;
	m_lastPositionY = y;
	m_lastPositionValid = true;

	state.value2d = Vec2{x, y};
	state.timestamp = timestamp;
	if (code == kPointerPosition) {
		m_inputControllerX = x;
		m_inputControllerY = y;
	}

	auto& delta = m_buttonStates[kPointerDelta];
	const bool moved = dx != 0.0f || dy != 0.0f;
	const bool wasPressed = delta.pressed;
	delta.value2d = Vec2{dx, dy};
	delta.value = std::hypot(dx, dy);
	delta.timestamp = timestamp;
	delta.justreleased = !moved && wasPressed;
	delta.pressed = moved;
	delta.justpressed = moved && !wasPressed;
	delta.waspressed = moved || wasPressed;
	delta.consumed = false;
}

void PointerInput::ingestAxis1(const std::string& code, f32 value, f64 timestamp) {
	auto& state = m_buttonStates[code];
	state.value = value;
	state.timestamp = timestamp;
	if (code != kPointerWheel || value == 0.0f) {
		if (code == kPointerWheel) {
			m_inputControllerWheel = value;
		}
		return;
	}
	m_inputControllerWheel = value;
	state.pressed = true;
	state.justpressed = true;
	state.justreleased = false;
	state.waspressed = true;
	state.consumed = false;
	state.pressedAtMs = timestamp;
	state.releasedAtMs = std::nullopt;
	state.pressId = m_nextPressId++;
}

void PointerInput::writeInputControllerPointerSnapshot(InputControllerSnapshot& snapshot) const {
	snapshot.pointerButtons |= m_inputControllerButtons;
	snapshot.pointerX = m_inputControllerX;
	snapshot.pointerY = m_inputControllerY;
	snapshot.pointerWheel = m_inputControllerWheel;
}

} // namespace bmsx

File: machine/ts/input/pointer.ts
import { machineManager } from '../core/machine_manager';
import { getPressedState, makeButtonState, resetObject } from './manager';
import type { ButtonState, InputHandler, KeyOrButtonId2ButtonState } from './models';
import type { VibrationParams } from '../platform';
import {
	type InputControllerPadSnapshot,
	INP_POINTER_BUTTON_AUX,
	INP_POINTER_BUTTON_BACK,
	INP_POINTER_BUTTON_FORWARD,
	INP_POINTER_BUTTON_PRIMARY,
	INP_POINTER_BUTTON_SECONDARY,
	type InputControllerSnapshot,
} from '../machine/devices/input/contracts';

const POINTER_DEFAULT_CODES = [
	'pointer_primary',
	'pointer_secondary',
	'pointer_aux',
	'pointer_back',
	'pointer_forward',
	'pointer_position',
	'pointer_delta',
	'pointer_wheel',
] as const;

function pointerButtonBit(code: string): number {
	switch (code) {
		case 'pointer_primary': return INP_POINTER_BUTTON_PRIMARY;
		case 'pointer_aux': return INP_POINTER_BUTTON_AUX;
		case 'pointer_secondary': return INP_POINTER_BUTTON_SECONDARY;
		case 'pointer_back': return INP_POINTER_BUTTON_BACK;
		case 'pointer_forward': return INP_POINTER_BUTTON_FORWARD;
		default: return -1;
	}
}

export class PointerInput implements InputHandler {
	public static readonly VIRTUAL_POINTER_INDEX = 0x7fffffff;
	public readonly gamepadIndex: number = PointerInput.VIRTUAL_POINTER_INDEX;

	private buttonStates: KeyOrButtonId2ButtonState = {};
	private nextPressId = 1;
	private lastPosition = { x: 0, y: 0 };
	private lastPositionValid = false;
	private lastDeltaTimestamp = 0;
	private lastWheelTimestamp = 0;
	private inputControllerButtons = 0;
	private inputControllerX = 0;
	private inputControllerY = 0;
	private inputControllerWheel = 0;

	constructor(public readonly deviceId: string = 'pointer:0') {
		this.reset();
	}

	public get supportsVibrationEffect(): boolean {
		return false;
	}

	public applyVibrationEffect(_params: VibrationParams): void { }

	public pollInput(): void {
		const now = machineManager.platform.clock.now();
		for (const key of Object.keys(this.buttonStates)) {
			const state = this.buttonStates[key];
			if (!state) continue;
			if (state.pressed) {
				const pressedAt = state.pressedAtMs ?? state.timestamp ?? now; // TODO: USE resolveStateTimestamp
				state.presstime = now - pressedAt;
				state.justpressed = false;
			} else {
				state.presstime = null;
				state.justreleased = false;
			}
			if (key === 'pointer_delta') {
				const ts = state.timestamp ?? 0;
				if (ts === this.lastDeltaTimestamp) {
					state.value2d = [0, 0];
					state.value = 0;
					state.pressed = false;
					state.justpressed = false;
					state.justreleased = false;
				} else {
					this.lastDeltaTimestamp = ts;
				}
			} else if (key === 'pointer_wheel') {
				const ts = state.timestamp ?? 0;
				if (ts === this.lastWheelTimestamp) {
					const wasPressed = state.pressed;
					state.value = 0;
					this.inputControllerWheel = 0;
					state.pressed = false;
					state.justpressed = false;
					state.justreleased = wasPressed;
				} else {
					this.lastWheelTimestamp = ts;
					state.justreleased = false;
				}
			}
			state.waspressed = state.waspressed || state.pressed;
			state.wasreleased = state.wasreleased || (!state.pressed);
			state.consumed = state.consumed ?? false;
		}
	}

	// disable-next-line single_line_method_pattern -- InputHandler state API exposes pointer-owned button storage through the shared button-state projection.
	public getButtonState(btn: string): ButtonState {
		return getPressedState(this.buttonStates, btn);
	}

	public writeInputControllerKeyWords(_keyWords: Uint32Array): void { }

	public writeInputControllerPointerSnapshot(snapshot: InputControllerSnapshot): void {
		snapshot.pointerButtons = (snapshot.pointerButtons | this.inputControllerButtons) >>> 0;
		snapshot.pointerX = this.inputControllerX;
		snapshot.pointerY = this.inputControllerY;
		snapshot.pointerWheel = this.inputControllerWheel;
	}

	public writeInputControllerPadSnapshot(_snapshot: InputControllerPadSnapshot): void { }

	private rebuildInputControllerState(): void {
		this.inputControllerButtons = 0;
		for (const code in this.buttonStates) {
			if (this.buttonStates[code].pressed) {
				const bit = pointerButtonBit(code);
				if (bit >= 0) {
					this.inputControllerButtons = (this.inputControllerButtons | (1 << bit)) >>> 0;
				}
			}
		}
		const position = this.buttonStates['pointer_position'];
		if (position && position.value2d) {
			this.inputControllerX = position.value2d[0];
			this.inputControllerY = position.value2d[1];
		} else {
			this.inputControllerX = 0;
			this.inputControllerY = 0;
		}
		const wheel = this.buttonStates['pointer_wheel'];
		this.inputControllerWheel = wheel && wheel.value !== undefined ? wheel.value : 0;
	}

	public ingestButton(code: string, state: ButtonState): void {
		const target = { ...state, value2d: state.value2d ? ([state.value2d[0], state.value2d[1]] as [number, number]) : null };
		if (target.pressed) {
			if (!target.pressId) target.pressId = this.nextPressId++;
			if (!target.pressedAtMs) target.pressedAtMs = target.timestamp ?? machineManager.platform.clock.now();
		} else {
			target.consumed = false;
		}
		this.buttonStates[code] = target;
		const bit = pointerButtonBit(code);
		if (bit >= 0) {
			const mask = 1 << bit;
			this.inputControllerButtons = target.pressed ? ((this.inputControllerButtons | mask) >>> 0) : ((this.inputControllerButtons & ~mask) >>> 0);
		}
	}

	public ingestAxis2(code: string, x: number, y: number, timestamp: number): void {
		const current = this.buttonStates[code] ?? makeButtonState();
		const dx = this.lastPositionValid ? (x - this.lastPosition.x) : 0;
		const dy = this.lastPositionValid ? (y - this.lastPosition.y) : 0;
		this.lastPosition.x = x;
		this.lastPosition.y = y;
		this.lastPositionValid = true;
		current.value2d = [x, y];
		current.timestamp = timestamp;
		this.buttonStates[code] = current;
		if (code === 'pointer_position') {
			this.inputControllerX = x;
			this.inputControllerY = y;
		}

		const delta = this.buttonStates['pointer_delta'] ?? makeButtonState();
		const moved = dx !== 0 || dy !== 0;
		const wasPressed = delta.pressed;
		delta.value2d = [dx, dy];
		delta.value = Math.hypot(dx, dy);
		delta.timestamp = timestamp;
		delta.justreleased = !moved && wasPressed;
		delta.pressed = moved;
		delta.justpressed = moved && !wasPressed;
		delta.waspressed = moved || wasPressed;
		delta.consumed = false;
		this.buttonStates['pointer_delta'] = delta;
	}

	public ingestAxis1(code: string, x: number, timestamp: number): void {
		const current = this.buttonStates[code] ?? makeButtonState();
		current.value = x;
		current.timestamp = timestamp;
		if (code === 'pointer_wheel') {
			this.inputControllerWheel = x;
			const hasDelta = x !== 0;
			if (hasDelta) {
				current.pressed = true;
				current.justpressed = true;
				current.justreleased = false;
				current.waspressed = true;
				current.consumed = false;
				current.pressedAtMs = current.timestamp;
				current.pressId = this.nextPressId++;
			}
		}
		this.buttonStates[code] = current;
	}

	public consumeButton(button: string): void {
		const state = this.buttonStates[button];
		if (state) {
			state.consumed = true;
			if (button === 'pointer_wheel') {
				state.pressed = false;
				state.justpressed = false;
				state.justreleased = false;
			}
		}
	}

	public reset(except?: string[]): void {
		if (!except) {
			this.buttonStates = {};
			for (const code of POINTER_DEFAULT_CODES) {
				this.buttonStates[code] = makeButtonState();
			}
			this.lastPosition = { x: 0, y: 0 };
			this.lastPositionValid = false;
			this.lastDeltaTimestamp = 0;
			this.lastWheelTimestamp = 0;
			this.inputControllerButtons = 0;
			this.inputControllerX = 0;
			this.inputControllerY = 0;
			this.inputControllerWheel = 0;
			return;
		}
		resetObject(this.buttonStates, except);
		this.rebuildInputControllerState();
	}
}
